Reconcile Triggers on Broker changes.

nachocano opened the pull request:
## Proposed Changes

- Watch for broker changes inside trigger controller and update it accordingly.
- Using a map to keep track of the triggers that should be updated for a particular broker.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE
```

nachocano committed:
Merge branch 'broker-new-model' into broker-new

nachocano committed:
Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Attempt to reconcile broker in trigger controller

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Deleting and re-creating subscription object as the backing channel
spec is immutable.

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Adding comment

nachocano commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -503,9 +596,17 @@ func (r *reconciler) subscribeToBrokerChannel(ctx context.Context, t *v1alpha1.T
+			logging.FromContext(ctx).Info("Cannot delete subscription", zap.Error(err))
+			return nil, err
+		}
+		sub = expected
```
@Harwayne please, take a special look at this. We are not updating the spec only, but rather the whole object.
If I do sub.spec = expected.spec, the client then complaints that we cannot create something with a revisionNumber. Tried setting the revision to the empty string, but still complaints.

Harwayne replied:
I think this is the correct thing to do. We don't need anything from the now deleted object, we just want to make a new one.

nachocano replied:
Sounds good!

Harwayne commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -137,27 +170,31 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 	} else {
 		logging.FromContext(ctx).Debug("Trigger reconciled")
-		r.recorder.Event(trigger, corev1.EventTypeNormal, triggerReconciled, "Trigger reconciled")
+		r.recorder.Eventf(trigger, corev1.EventTypeNormal, triggerReconciled, "Trigger reconciled: %q", trigger.Name)
```
Is the name already there in the sense that the source of the event is the trigger?

nachocano replied:
I added it so that an event line has all the info.
Example:
Normal   TriggerReconciled       2s (x15 over 6h32m)   trigger-controller  Trigger reconciled: "t"

I can remove the name if you prefer

Harwayne replied:
How did you get that event? If I describe the Trigger, then it implicitly means that the event ties to that Trigger (so the Trigger's name is redundant, as the name is identical to the source). If I list events, then I can look at the source field, which has the trigger's name already.

nachocano replied:
Yes, you are right... I saw that you merged it in... We won't add that info then in the future, is redundant.

Harwayne commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -503,9 +596,17 @@ func (r *reconciler) subscribeToBrokerChannel(ctx context.Context, t *v1alpha1.T
+		// We delete it instead, and re-create it.
+		err = r.client.Delete(ctx, sub)
+		if err != nil {
+			logging.FromContext(ctx).Info("Cannot delete subscription", zap.Error(err))
```
Ideally we would emit an event here too (both if the delete works or fails). Doesn't need to be in this PR.

nachocano replied:
I added it in the case it fails as we were doing in most of the other controllers. The only succeed event we emit is when it reconciles.
With this change we will have both SubscriptionDeleteFailed and TriggerReconcileFailed... On future PRs, we can add more detailed events for all of the other possible errors, and remove the redundant TriggerReconcileFailed...
Let me know if you agree, otherwise I can rollback...

Harwayne commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -503,9 +596,17 @@ func (r *reconciler) subscribeToBrokerChannel(ctx context.Context, t *v1alpha1.T
+		sub = expected
+		err = r.client.Create(ctx, sub)
 		if err != nil {
+			logging.FromContext(ctx).Info("Cannot create subscription", zap.Error(err))
```
Ideally we would emit an event here too (both if the create works or fails). Doesn't need to be in this PR.

nachocano replied:
Same as above

Harwayne commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -67,19 +74,23 @@ type reconciler struct {
 	recorder      record.EventRecorder
 
+	triggersLock sync.RWMutex
+	triggers     map[string]map[reconcile.Request]bool
```
I think we can make the key the namespaced name of the Broker, that way we only reconcile triggers that correspond to the specific broker.

`map[types.NamespacedName]map[reconcile.Request]bool`

And possibly swap out bool for struct{}.

nachocano replied:
Done

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into broker-new

nachocano committed:
Adding event messages

nachocano committed:
Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

nachocano commented on pkg/apis/eventing/v1alpha1/trigger_validation.go:
```diff
@@ -38,7 +38,7 @@ func (ts *TriggerSpec) Validate() *apis.FieldError {
 
 	if ts.Filter.SourceAndType == nil {
-		fe := apis.ErrMissingField("filter.exactMatch")
+		fe := apis.ErrMissingField("filter.sourceAndType")
```
included it as part of this PR, although it has nothing to do :)
very cosmetic change

Harwayne replied:
You rock!

nachocano commented:
@Harwayne ready to review

Harwayne commented on pkg/reconciler/v1alpha1/trigger/trigger.go:
```diff
@@ -66,16 +66,23 @@ const (
-	triggers     map[string]map[reconcile.Request]bool
+	// Contains the triggers that correspond to a particular broker.
+	// We use this to reconcile only the triggers that correspond to a certain broker.
+	// brokerNamespacedName -> triggerReconcileRequest -> dummy struct
```
Awesome comment!
